### DISCUSSION TITLE
How to track state change triggered by Alexa?

### DISCUSSION THREAD

**Discussion by giddyhup** [arendst/Tasmota#18285] [Q&A]
My Tasmota-enabled light switch can be triggered via three ways: a touch sensor, via MQTT, or via Alexa. 

MQTT is used in connection with a Node RED flow which acts on motion sensor events. The flow turns the light off again after some time has passed. When the light is turned on via touch sensor or through an Alexa command it should stay on (even when a motion sensor event would have eventually led to an OFF command). In Node RED set a 'latched' variable for this purpose.

Regrettably, keeping track of the triggers outside of the motion sensor is cumbersome. Here's how I am currently handling it. I keep track of the light state (ON/OFF) by MQTT subscription. I redirected the button to no longer act locally but to publish to a topic so that I can store the trigger (and then turn the actual light on/off via MQTT). When I receive a light state change message and there was no button event previous to it then I assume the trigger was Alexa and store that.

**Is there a better way to find out if the light state change was triggered by Alexa? Would it be possible to provide the information via MQTT?**

**Comment by Jason2866** [comment]
Every change of a state generates a mqtt message. The source of the actor is not relevant.
So a change initiated from Alexa will generate a mqtt message

**Reply by giddyhup** [reply]
Thank you for your reply. In some cases such as mine the initiator is relevant and therefore "state changed" is less helpful than "state changed by".

**Reply by Jason2866** [reply]
Not implemented and will not be done.

**Comment by barbudor** [comment]
@giddyhup This is where generaly rules and event comes in

I don't know how Alexa comes into play but you can have all 3 sources managed through EVENT command instead of direct POWER command and use rules and rulestimer to manage your state machine
Not having spent more than 30 seconds to think about it but let assume you get Alexa to sent command to event ALEXA and the moion sensor sending to event MOTION, you may get a start with
```
Rule1
  on event#ALEXA=ON do backlog var1 1; POWER ON endon
  on event#ALEXA=OFF do backlog var1 0; POWER OFF endon
  on event#MOTION=1 do backlog POWER ON; RuleTimer1 30 endon
  on Rules#Timer=1 do POWER %var1% endon
```
Button might be tricky. So far it may be necessary to define a virtual realy associated to the touch button

But that was to give you some idea, not a cooked solution

**Reply by giddyhup** [reply]
Thank you for the constructive suggestion. I'll look into triggers. 

If there is an Alexa trigger event that would already help me. I could send an MQTT message when it happens. [This discussion](https://github.com/arendst/Tasmota/discussions/13763) mentions a complete list of triggers. Regrettably, I haven't found said list in general or specifically a hint about an Alexa trigger.

**Reply by sfromis** [reply]
Do not look for a "complete list of triggers", that cannot exist. Instead look in the Tasmota console, if you have a message tagged `MQT:` or if not connected, `RSL:`, with a JSON payload, that is where you have your triggers, which you can use in rules.

**Reply by giddyhup** [reply]
Thanks for the helpful pointer.

I had to set the log level to 4 and then I saw this: `HTP: WeMo basic event from 192.168.XXX.XXX`. I guess that's what I am looking for and I also hope there are no false positives (like device scans). 

The message is not JSON-formatted, how would I use this in a rule?

**Reply by sfromis** [reply]
Well, I specifically told you what you *can* use in rules. Other messages are not available to rules, as no triggers are generated for "any message", only the JSON payloads potentially published via MQTT.

**Reply by s-hadinger** [reply]
I think it was already reported that Alexa triggered events are not able to trigger rules. I will look at what's missing

**Reply by giddyhup** [reply]
@sfromis I didn't want to appear stubborn. Sorry if it seemed that way. I just tried to put @barbudor's suggestion and your input together. I guess I hoped that the JSON format wasn't a hard requirement.
